LocalFileSource

ivovandongen opened the pull request:
Fixes #6273

ivovandongen commented:
@tmpsantos Could you review please (assuming the ci completes)?

I'm not very happy with the name `FileFileSource`, but couldn't come up with a better one.

1ec5 commented:
> I'm not very happy with the name `FileFileSource`, but couldn't come up with a better one.

How about `LocalFileSource`?

1ec5 commented:
With these changes, is it still necessary for the default `AssetFileSource` to know how to handle absolute paths? Also, [this iOS/macOS special case](https://github.com/mapbox/mapbox-gl-native/blob/1777b8757a437d6f27928c2bb3d821fc8679cf20/platform/darwin/src/NSURL%2BMGLAdditions.m#L10-L17) can be removed now that file: URLs are supported in core.

brunoabinader commented:
> How about LocalFileSource?

I like this one - though we should be aware that the [file URI scheme](https://en.wikipedia.org/wiki/File_URI_scheme) can also point to files on a network location.

tmpsantos commented:
> How about LocalFileSource?

:+1:

ivovandongen commented:
> How about LocalFileSource?

Seems popular :) I've changed this.

> is it still necessary for the default AssetFileSource to know how to handle absolute paths? 

I assumed we need to leave it in there now for backwards compatibility. 

> this iOS/macOS special case can be removed now that file: URLs are supported in core.

Removed

tmpsantos commented on platform/default/default_file_source.cpp:
```diff
@@ -19,6 +20,12 @@ bool isAssetURL(const std::string& url) {
     return std::equal(assetProtocol.begin(), assetProtocol.end(), url.begin());
 }
 
+const std::string fileProtocol = "file://";
```
Better avoid this. See for alternatives: https://github.com/mapbox/mapbox-gl-native/pull/6455

tmpsantos commented on test/storage/local_file_source.test.cpp:
```diff
@@ -0,0 +1,101 @@
+
+namespace {
+
+std::string toAbsoluteUrl(const std::string& fileName) {
```
Nit: we use URL on the internal bits rather then Url.

tmpsantos commented on test/storage/local_file_source.test.cpp:
```diff
@@ -0,0 +1,101 @@
+std::string toAbsoluteUrl(const std::string& fileName) {
+    char buff[PATH_MAX + 1];
+    char* cwd = getcwd( buff, PATH_MAX + 1 );
+    return "file:://" + std::string(cwd) + "/test/fixtures/storage/assets/" + fileName;
```
`://`

tmpsantos commented on platform/default/local_file_source.cpp:
```diff
@@ -0,0 +1,55 @@
+public:
+    void request(const std::string& url, FileSource::Callback callback) {
+        //Cut off the protocol
+        std::string path = mbgl::util::percentDecode(url.substr(7));
```
Make this like #6455 with the size a global together with the scheme variable.

tmpsantos commented on test/storage/local_file_source.test.cpp:
```diff
@@ -0,0 +1,101 @@
+    loop.run();
+}
+
+TEST(LocalFileSource, URLEncoding) {
```
For the sake we are no missing anything, can you make a test with a extremely large file name, like `memset(filename, 'x', PATH_MAX);`

Wondering what would happen if you try to read something you don't have permission to (like `/root/foo.txt`).

ivovandongen commented:
@tmpsantos Thanks for the review. I've cleared up all items.

>Wondering what would happen if you try to read something you don't have permission to 

It returns `Response::Error::Reason::Other`. Not quite sure how to test this in a platform independent way though.

tmpsantos commented:
Core part looks fine to me. /cc @tobrun for the Android bits.

tmpsantos commented on platform/default/local_file_source.cpp:
```diff
@@ -0,0 +1,63 @@
+
+namespace mbgl {
+    
+const char* protocol = "file://";
```
This could go to the anonymous namespace.

ivovandongen replied:
Thanks! Changed it.

tmpsantos commented on platform/default/local_file_source.cpp:
```diff
@@ -0,0 +1,62 @@
+#include <sys/stat.h>
+#include <unistd.h>
+
+const char* protocol = "file://";
```
```
namespace {

const char* protocol = "file://";
const std::size_t protocolLength = 7;

} // namespace
```

1ec5 commented:
>> is it still necessary for the default AssetFileSource to know how to handle absolute paths?

> I assumed we need to leave it in there now for backwards compatibility. 

True, removing that support might (?) break asset: assets that have been cached, I suppose, and it’s pretty harmless.

ivovandongen commented:
@tmpsantos I fixed the valgrind errors. If all else is good, could you press the review button?

tobrun commented:
:shipit:

bigsassy commented:
I just wanted to say to @ivovandongen, thank you!  Oh, and also @1ec5 and the rest of the mapbox team as well 👍

zugaldia commented:
cc: @cammace this looks like a good candidate for a new sample, or a required update to https://www.mapbox.com/android-sdk/examples/custom-raster/.

SeanChristopherConway commented:
I am using ios SDK 3.4.0-beta.6 and am not able to use the file:// style to load my style JSON or in the tiles section of my sources in the style.json.  When I use asset:// for the style it loads, but neither asset nor file seem to work within the styles section.

1ec5 commented:
It should be possible to use an _absolute_ file: URL as an MGLMapView’s style URL. But I’m not sure if it’s supposed to be possible to use a file: URL inside the JSON file itself. In any event, this closed PR likely isn’t being watched by folks who know the answer for sure, so please open a new issue, ideally with more details about your style JSON.
